File: day4/day4.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../utils/file.h"
#include "../utils/log.h"

typedef struct {
  unsigned char *value;
  int size;
} Map;

static inline int index_2d(int stride, int x, int y) { return y * stride + x; }

static int dfs(Map *m, int x, int y) {
  const int size = m->size;
  const int stride = size + 1;
  unsigned char *restrict grid = m->value;

  const int y_start = (x >= 0 && y > 0) ? y - 1 : 0;
  const int y_end = (y + 1 < size) ? y + 1 : (size - 1);
  const int x_start = (x > 0) ? x - 1 : 0;
  const int x_end = (x + 1 < size) ? x + 1 : (size - 1);

  int neighbor_count = 0;
  for (int cy = y_start; cy <= y_end; ++cy) {
    unsigned char *row = grid + cy * stride + x_start;
    for (int cx = x_start; cx <= x_end; ++cx, ++row) {
      if (cy == y && cx == x) {
        continue;
      }
      if (*row == '@') {
        if (++neighbor_count > 3) {
          return 0;
        }
      }
    }
  }

  grid[index_2d(stride, x, y)] = 'x';

  int result = 1;
  for (int cy = y_start; cy <= y_end; ++cy) {
    unsigned char *row = grid + cy * stride + x_start;
    for (int cx = x_start; cx <= x_end; ++cx, ++row) {
      if (*row == '@') {
        result += dfs(m, cx, cy);
      }
    }
  }

  return result;
}

int main(int argc, char **argv) {
  if (argc < 2) {
    fprintf(stderr, "Usage: %s <file_input>\n", argv[0]);
    return 1;
  }

  unsigned char *data = NULL;
  size_t data_size = 0;
  if (!read_entire_file(argv[1], &data, &data_size)) {
    fprintf(stderr, "Error opening file\n");
    return 1;
  }

  Map m;
  m.value = data;
  m.size = 0;

  for (size_t i = 0; i < data_size; ++i) {
    unsigned char c = data[i];
    if (c == '\r') {
      fprintf(stderr, "Remove CR from data\n");
      free(data);
      return 1;
    }
    if (c == '\n') {
      break;
    }
    m.size += 1;
  }

  log("m.size = %d\n", m.size);

  const int stride = m.size + 1;
  int counter = 0;

  for (int y = 0; y < m.size; ++y) {
    unsigned char *row = m.value + y * stride;
    for (int x = 0; x < m.size; ++x) {
      if (row[x] == '@') {
        counter += dfs(&m, x, y);
      }
    }
  }

#if LOGGER_ENABLED
  fwrite(m.value, 1, data_size, stdout);
  putchar('\n');
#endif

  print_value(counter, "%d");
  free(data);
  return 0;
}

File: day4/day4.rs
use advent2025::debug_println;
use std::{env, fs, io};

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    let Some(file_path) = args.get(1) else {
        eprintln!(
            "Usage: {} <file_input>",
            args.first().map(|s| s.as_str()).unwrap_or("day4")
        );
        return Err(io::Error::other("file_input missing"));
    };

    let mut data = fs::read(file_path)?;

    let mut size = 0usize;
    for &b in &data {
        if b == b'\r' {
            eprintln!("Remove CR from data");
            return Err(io::Error::other("input has CR"));
        }
        if b == b'\n' {
            break;
        }
        size += 1;
    }

    debug_println!("m.size = {}", size);

    let stride = size + 1;
    let mut counter: i32 = 0;

    for y in 0..size {
        let row_base = y * stride;
        for x in 0..size {
            let idx = row_base + x;
            if data[idx] == b'@' {
                counter += dfs(&mut data, size, stride, x, y);
            }
        }
    }

    // Optional: dump final map in debug builds
    debug_println!("{}", String::from_utf8_lossy(&data));
    println!("counter = {counter}");

    Ok(())
}

fn dfs(map: &mut [u8], size: usize, stride: usize, x: usize, y: usize) -> i32 {
    let y_start = y.saturating_sub(1);
    let y_end = (y + 1).min(size - 1);
    let x_start = x.saturating_sub(1);
    let x_end = (x + 1).min(size - 1);

    let mut neighbor_count = 0;
    for cy in y_start..=y_end {
        let mut idx = cy * stride + x_start;
        for cx in x_start..=x_end {
            if cy == y && cx == x {
                idx += 1;
                continue;
            }
            if map[idx] == b'@' {
                neighbor_count += 1;
                if neighbor_count > 3 {
                    return 0;
                }
            }
            idx += 1;
        }
    }

    map[y * stride + x] = b'x';

    let mut result = 1;
    for cy in y_start..=y_end {
        let mut idx = cy * stride + x_start;
        for cx in x_start..=x_end {
            if map[idx] == b'@' {
                result += dfs(map, size, stride, cx, cy);
            }
            idx += 1;
        }
    }

    result
}
